presto-main/src/main/java/com/facebook/presto/execution/buffer/BroadcastOutputBuffer.java

```diff
@@ -90,6 +87,9 @@ public BroadcastOutputBuffer(
                 requireNonNull(maxBufferSize, "maxBufferSize is null").toBytes(),
                 requireNonNull(systemMemoryContextSupplier, "systemMemoryContextSupplier is null"),
                 requireNonNull(notificationExecutor, "notificationExecutor is null"));
+        this.pageTracker = new LifespanSerializedPageTracker(memoryManager, Optional.of((lifespan, releasedPageCount, releasedSizeInBytes) -> {
+            checkState(totalBufferedPages.addAndGet(-releasedPageCount) >= 0);
```

Comment:
this check looks quite universal -- wondering why it cannot be used for other buffer as well?

Reply:
See above comments, but I'm not sure the extra cost of tracking the count makes sense there since the number of buffered pages is more easily reasoned about through other means except for the broadcast buffer.